### ISSUE TITLE
Use semantic constructs for sections

### ISSUE THREAD

**Issue by Mogztter** [asciidoctor/asciidoctor#3981]
This issue will be addressed as part of: https://github.com/asciidoctor/asciidoctor/issues/242

### Sections

I think we should use [`<section>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section) as a parent element.
I can't think of any other element that would make more sense.

```html
<section>
  <!-- ... -->
</section>
```

#### Headings

Here's MDN usage recommandation:

> To reiterate, each `<section>` should be identified, typically by including a heading (`<h1>`-`<h6>` element) as a child of the `<section>` element, wherever possible.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section#usage_notes

This recommandation fits quite well with the AsciiDoc syntax:

```
== Section Title

This is a paragraph.
```

```html
<section>
  <h2>Section Title</h2>
  <p>This is a paragraph.</p>
</section>
```

#### Wrapping content

I don't think we need to wrap the content of a section inside a generic container such as the `<div>` element:

```html
<section>
  <h2>Section Title</h2>
  <p>First paragraph.</p>
  <p>Second paragraph.</p>
</section>
```

#### Nesting sections

I did some research and I think it's fine to nest section elements:

```html
<section>
  <h2>Section Title</h2>
  <section>
    <h3>Subsection Title</h3>
  </section>
</section>
```

##### Using CSS

```css
body {
  counter-reset : h2;
}

h2 {
  counter-reset : h3;
}

h3 {
  counter-reset : h4;
}

h4 {
  counter-reset : h5;
}

h5 {
  counter-reset : h6;
}

h2:before {
  content : counter(h2,decimal) ".\0000a0";
  counter-increment : h2;
}

h3:before {
  content : counter(h2,decimal) "." counter(h3,decimal) ".\0000a0";
  counter-increment : h3;
}

h4:before {
  content : counter(h2,decimal) "." counter(h3,decimal) "." counter(h4,decimal) ".\0000a0";
  counter-increment : h4;
}
```

Note: `\0000a0` is a non-breaking space.

One benefit is that the `counter()` CSS function takes a counter style name as the second parameter.
This can be useful when you want to mix counter styles.
Here, I want to use roman numerals on chapters, decimal numbering on sections and lower alpha on subsections:

If we decide to use CSS then we will need to dynamically generate CSS when `sectnums` attribute is defined and honor `part-signifier`, `chapter-signifier`, `sectnumlevels`.

Related issue: https://github.com/asciidoctor/asciidoctor/issues/3311

##### Alternatives

We could keep using the `Section#sectnum` method in the converter to prefix the section title.

#### Section level

Should we add a role or an attribute with the current level of a section?
The goal would be to make it easier to style using CSS.
A data attribute might be useful if we want to manipulate/use this value from JavaScript:

```html
<section data-level="1">
  <h2>Section Title</h2>
  <p>This is a paragraph.</p>
</section>
```

#### Section level 0

A document can only have multiple level 0 sections if the `doctype` is set to `book`.
Currently, the HTML converter will use an `<h1>` element.

MDN clearly states that using more than one `<h1>` is not recommended:

> Use only one `<h1>` per page or view. It should concisely describe the overall purpose of the content.
> Using more than one `<h1>` will not result in an error, but is not considered a best practice. Using only one `<h1>` is beneficial for screenreader users, and SEO.
> While HTML5 allows a `<h1>` per sectioning element, it is not considered best practice, and may subvert the expectations of how screen reader users navigate.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/heading_elements#usage_notes

Maybe we should use an `<h2>` with a special role or attribute to distinguish them from section level 1?
I can't think of any good name to describe a section level 0 (apart from `section-0` 😅).

Another idea would be to use a data attribute:

```html
<section data-level="0">
  <h2>Dedication</h2>
</section>
```

Please note that we can use the data attribute as a CSS selector:

```css
section[data-level="0"] > h2:before {
  content : "";
  counter-increment : none;
}

section[data-level="0"] > h2 {
  color: deeppink;
}
```

#### Roles and id

Here's the recommandation from https://developers.google.com:

> To add an anchor to a heading in HTML, do the following:
> Add a <section> element with an id attribute. Don't use <a name>.
> Use lowercase for id values, and put hyphens between words.

https://developers.google.com/style/headings-targets#adding-an-anchor

```html
<section id="introduction-to-everything">
  <h2>Introduction to everything</h2>
  <!-- ... -->
</section>
```

I think we should this recommandation and assign roles and id to the `<section>` element:

```
[#release-policies.primary]
== Version and Release Policies

This is a paragraph.
```

```html
<section id="release-policies" class="primary">
  <h2>Version and Release Policies</h2>
  <p>This is a paragraph.</p>
</section>
```

**Comment by mojavelinux** [comment]
Thank you for another great proposal! I'm really excited to take this next big step with this converter.

I absolutely agree that `<section>` is the right tag here. I also agree that the heading tags should match the section depth. That's a great start.

I'm not convinced we should rely entirely on CSS for counting. That strips some of the semantics out of the HTML. We could consider a mode that delegates section numbering to CSS, but I would consider that to be a specialized use case. And it could result in certain features, like xrefstyle, being disabled. What we do probably want to do is wrap the section number in a span so that it's appearance and flow can be controlled independently of the title text.

I don't think we should shy away from using h1 for part titles. Although using multiple h1 tags is not a recommended practice for a single article, the book doctype is meant to accommodate multiple parts, which are essentially multiple discrete "pages". Hence why the h1 appearing multiple times makes sense in this case. One possible idea to make this more semantic would be to make each part are separate article (`<article class="part">`). However, I haven't explored how that would work in practice.

I think you're right that ID and roles need to go on the section tag. That's because, currently, it's not possible to style a parent element using CSS. So if you we don't put them on the section, there is no way to control the section itself using the ID or role.

One idea is to duplicate the ID and roles on the section with a special prefix, like sect- (perhaps configurable). That would make it easy to style the heading tag, but still leave the door open for styling the section itself. Something to consider, perhaps.

**Comment by Mogztter** [comment]
> I'm not convinced we should rely entirely on CSS for counting. That strips some of the semantics out of the HTML. We could consider a mode that delegates section numbering to CSS, but I would consider that to be a specialized use case. And it could result in certain features, like xrefstyle, being disabled. 

I just realized that we don't have to do anything special to support CSS numbering. It's already possible to add section numberin using a custom style (as shown above).

> What we do probably want to do is wrap the section number in a span so that it's appearance and flow can be controlled independently of the title text.

That's a good idea 👍🏻 

> I don't think we should shy away from using h1 for part titles. Although using multiple h1 tags is not a recommended practice for a single article, the book doctype is meant to accommodate multiple parts, which are essentially multiple discrete "pages". Hence why the h1 appearing multiple times makes sense in this case. One possible idea to make this more semantic would be to make each part are separate article (<article class="part">). However, I haven't explored how that would work in practice.

I've decided to address this concern in a separate issue.
The book doctype is a bit specific and, as you've mentioned, we will probably need new semantic elements.

> I think you're right that ID and roles need to go on the section tag. That's because, currently, it's not possible to style a parent element using CSS. So if you we don't put them on the section, there is no way to control the section itself using the ID or role.

👍🏻 

> One idea is to duplicate the ID and roles on the section with a special prefix, like sect- (perhaps configurable). That would make it easy to style the heading tag, but still leave the door open for styling the section itself. Something to consider, perhaps.

I don't think we need to do that for styling purposes.
We can style the section and/or the heading tag (using the child element):

```css
/* section style */
#_introduction {
  border: 1px solid #ccc;
  padding: 0.5rem 0 1rem;
}

/* section heading style */
#_introduction > h2 {
  font-weight: bold;
}
```

**Comment by mojavelinux** [comment]
> I don't think we need to do that for styling purposes.

I understand it's possible. Rather, I was addressing a common compliant that we get from designers that having to do it that way is cumbersome. By automatically adding a corresponding ID/class on the heading tag, it makes it easier for those designers to target the tag directly. It's an olive branch.

**Comment by Mogztter** [comment]
> I understand it's possible. Rather, I was addressing a common compliant that we get from designers that having to do it that way is cumbersome. By automatically adding a corresponding ID/class on the heading tag, it makes it easier for those designers to target the tag directly. It's an olive branch.

I see but everything has its drawbacks. I guess some users will not be pleased because it will generate additional IDs/classes (eventually producing conflicts with existing styles/frameworks/components).
Right now, I don't really see the benefit of adding corresponding ID/class on heading tags but someone could probably convince me otherwise...

Anyway, I think we should probably address this issue later since we haven't started working on styles yet. Once we get there, it would be useful to get designers (and more broadly people will solid understanding of CSS) involved. We could ask them to design a theme from the HTML produced by the new HTML converter in order to collect their feedback and pain points.

**Comment by mojavelinux** [comment]
> Anyway, I think we should probably address this issue later since we haven't started working on styles yet.

I agree. I was only bringing it up for thoroughness...so we keep it in mind. Probably best to address it later on in the process.
